src/System.Text.Encodings.Web/tests/JavaScriptStringEncoderTests.cs

```diff
@@ -239,6 +258,83 @@ public void JavaScriptStringEncode_BadSurrogates_ReturnsUnicodeReplacementChar()
             Assert.Equal(expected, retVal);
         }
 
+        [Fact]
+        public void JavaScriptEncoder_BadSurrogates_ReturnsUnicodeReplacementChar()
+        {
+            // Arrange
+            JavaScriptEncoder encoder = JavaScriptEncoder.Create(UnicodeRanges.All); // allow all codepoints
+
+            // "a<unpaired leading>b<unpaired trailing>c<trailing before leading>d<unpaired trailing><valid>e<high at end of string>"
+            const string input = "a\uD800b\uDFFFc\uDFFF\uD800d\uDFFF\uD800\uDFFFe\uD800";
+            const string expected = "a\uFFFDb\uFFFDc\uFFFD\uFFFDd\uFFFD\\uD800\\uDFFFe\uFFFD"; // 'D800' 'DFFF' was preserved since it's valid
+
+            // String-based Encode()
+            string retVal = encoder.Encode(input);
+            Assert.Equal(expected, retVal);
+
+            // OperationStatus-based Encode()
+            Span<char> destination = new char[23];
+            OperationStatus status = encoder.Encode(input.AsSpan(), destination, out int charsConsumed, out int charsWritten, isFinalBlock: true);
+            Assert.Equal(OperationStatus.Done, status);
+            Assert.Equal(13, charsConsumed);
+            Assert.Equal(13, input.Length);
+            Assert.Equal(23, charsWritten);
+            Assert.Equal(expected, new string(destination.Slice(0, charsWritten).ToArray()));
+        }
+
+        [Fact]
+        public void JavaScriptEncoder_BadSurrogates_NeedsMoreData()
```

Comment:
What happens if the input/output spans overlap? Might be worth adding a test for that.

Reply:
I believe it is undefined; the API doc states this must not happen. Do we guard or test against this in any other case?

Reply:
We recently updated `MemoryExtensions.Trim`, to explicitly guard against that and throw. Something to consider - however I don't think it is absolutely necessary. The trim case was different because we had incorrectly documented that it does support overlapping (where we are explicitly documenting that it doesn't).
https://github.com/dotnet/corefx/issues/34799

Alternatively, we could consider adding a guideline that all "OperationStatus" style APIs that take input/output spans should either document whether overlapping is supported or not (since they are generally for advance scenarios to begin with and perf sensitive), OR guard against overlaps explicitly in all cases.

cc @jkotas, @KrzysztofCwalina - what are your thoughts on that? Is documentation sufficient?

Reply:
>  what are your thoughts on that? Is documentation sufficient?

I think so. The existing `EncodeUtf8` method on this type is on this plan already.
`